File: measurements/test_py_bench/src/pmpymax_test.py
import concore
import time
import os
import psutil
import sys

# --- Script Configuration ---
concore.delay = 0.01
concore.default_maxtime(1000)   # must match cpymax's maxtime

init_simtime_u  = "[0.0, 0.0, 0.0]"
init_simtime_ym = "[0.0, 0.0, 0.0]"

# --- Measurement Initialization ---
process = psutil.Process(os.getpid())
start_time = time.monotonic()
message_count = 0
total_bytes = 0

# --- Main Script Logic ---
ym = concore.initval(init_simtime_ym)

print("pmpymax_test.py started...")

# pmpymax drives the pipeline: it writes first with delta=1 to kick off simtime
# Then on each tick it reads u, increments, and writes ym (advancing simtime)
while concore.simtime < concore.maxtime:
    while concore.unchanged():
        u = concore.read(1, "u", init_simtime_u)

    message_count += 1
    total_bytes += sys.getsizeof(u)

    ym[0] = u[0] + 10
    print(f"pmpymax: u={u[0]:.2f} -> ym={ym[0]:.2f}")

    concore.write(1, "ym", ym, delta=1)   # delta=1 advances simtime → drives the loop

# --- Finalize and Report ---
end_time = time.monotonic()
duration = end_time - start_time
cpu_usage = process.cpu_percent() / duration if duration > 0 else 0

print("\n" + "="*30)
print("--- PMPYMAX_TEST: FINAL RESULTS ---")
print(f"Total messages processed: {message_count}")
print(f"Total data processed:     {total_bytes / 1024:.4f} KB")
print(f"Total execution time:     {duration:.4f} seconds")
print(f"Approximate CPU usage:    {cpu_usage:.2f}%")
print(f"concore retry count:      {concore.retrycount}")
print("="*30)

File: measurements/pmpymax_test.jl
# pmpymax_test.jl
# Julia port of pmpymax_test.py
# Plant-model node: reads u from port 1, computes ym[1] = u[1] + 10, writes ym back.
# Drives simtime forward with delta=1 on each write.
import Concore: initval, concore_read, concore_write, unchanged, state, default_maxtime

function main()
    # --- Script Configuration ---
    state.delay = 0.01
    default_maxtime(1000.0)   # must match cpymax's maxtime

    init_simtime_u  = "[0.0, 0.0, 0.0]"
    init_simtime_ym = "[0.0, 0.0, 0.0]"

    # --- Measurement Initialization ---
    start_time    = time()
    message_count = 0
    total_bytes   = 0

    # --- Main Script Logic ---
    ym = initval(init_simtime_ym)

    println("pmpymax_test.jl started...")

    u = concore_read(1, "u", init_simtime_u)
    # pmpymax drives the pipeline: on each tick it reads u, increments, and writes ym
    # (advancing simtime with delta=1)
    while state.simtime < state.maxtime
        while unchanged()
            u = concore_read(1, "u", init_simtime_u)
        end

        message_count += 1
        total_bytes   += Base.summarysize(u)

        ym[1] = u[1] + 10.0
        println("pmpymax: u=$(round(u[1], digits=2)) -> ym=$(round(ym[1], digits=2))")

        concore_write(1, "ym", ym, 1)   # delta=1 advances simtime → drives the loop
    end

    # --- Finalize and Report ---
    end_time = time()
    duration = end_time - start_time

    println("\n" * "="^30)
    println("--- PMPYMAX_TEST: FINAL RESULTS ---")
    println("Total messages processed: $message_count")
    println("Total data processed:     $(round(total_bytes / 1024, digits=4)) KB")
    println("Total execution time:     $(round(duration, digits=4)) seconds")
    # CPU usage requires shelling out in Julia; omitted (no psutil equivalent in stdlib)
    println("concore retry count:      $(state.retrycount)")
    println("="^30)
end

main()
